src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorTextReplacer.cs

```diff
@@ -34,36 +34,57 @@ public RazorTextReplacer(ILogger<RazorTextReplacer> logger)
-                // include line offsets.
-                var documentLines = File.ReadAllLines(replacementGroup.Key);
-                var documentText = new StringBuilder();
-                foreach (var line in documentLines)
-                {
-                    documentText.AppendLine(line);
-                }
+                var documentTextStr = File.ReadAllText(replacementGroup.Key);
+                var lineOffsets = GetLineOffsets(documentTextStr).ToArray();
+                var documentText = new StringBuilder(documentTextStr);
 
                 foreach (var replacement in replacementGroup)
                 {
                     _logger.LogInformation("Updating source code in Razor document {FilePath} at line {Line}", replacement.FilePath, replacement.StartingLine);
 
+                    // If the original text doesn't fit in the lines of the original document, then the replacement is invalid
+                    if (replacement.StartingLine + replacement.OriginalText.Lines.Count >= lineOffsets.Length)
+                    {
+                        continue;
+                    }
+
                     // Start looking for replacements at the start of the indicated line
-                    var startOffset = GetLineOffset(documentLines, replacement.StartingLine);
+                    var startOffset = lineOffsets[replacement.StartingLine];
 
                     // Stop looking for replacements at the start of the first line after the indicated line plus the number of lines in the indicated text
-                    var endOffset = GetLineOffset(documentLines, replacement.StartingLine + replacement.OriginalText.Lines.Count);
+                    var endOffset = lineOffsets[replacement.StartingLine + replacement.OriginalText.Lines.Count];
 
                     // Trim the string that's being replaced because code from Razor code blocks will include a couple extra spaces (to make room for @{)
                     // compared to the source that actually appeared in the cshtml file.
                     var originalText = replacement.OriginalText.ToString().TrimStart();
                     var updatedText = replacement.NewText.ToString().TrimStart();
-                    MinimizeReplacement(ref originalText, ref updatedText);
+
+                    // Generally, it's not necessary to minimize replacements since the text should be the same in both documents
+                    // However, in the specific case of text being added before the first #line pragma in a Razor doc, it's possible
+                    // that C# unrelated to the original document will show up with the replaced text, so minimize replacements only
+                    // in the case that text is being inserted before the original first line.
+                    if (replacement.StartingLine == 0)
+                    {
+                        MinimizeReplacement(ref originalText, ref updatedText);
+                    }
```

Comment:
I'm still not entirely happy with how we minimize replacements. I'm going to re-work this but will do it in a separate PR since this is focused on implementing tests.